macros/contexts/contextBoolean.pl

```diff
@@ -0,0 +1,253 @@
+		'+'  => { alias => 'or' },
+		'!'  => { alias => 'not' },
+		'u-' => { alias => 'not' },
+		'><' => { alias => 'xor' },
+	);
+
+	## Set default variables 'p' and 'q'
+	$context->variables->are(
+		p => { isValue => 1, type => Value::Type('Boolean') },
+		q => { isValue => 1, type => Value::Type('Boolean') },
+	);
+}
+
+## Subclass Value::Formula for boolean formulas
+package context::Boolean::Formula;
+our @ISA = ('Value::Formula');
+
+sub new {
+	my $class = shift;
+	my $self  = $class->SUPER::new(@_);
+
+	my @variables = $self->{context}->variables->names;
+	my @points;
+	my @values;
+
+	my $f = $self->{f} = $self->perlFunction(undef, \@variables);
+
+	foreach my $combination (0 .. 2**@variables - 1) {
+		my @point = map { $combination & 2**$_ ? 1 : 0 } (0 .. $#variables);
+		my $value = &$f(@point) ? 1 : 0;
+		push @points, \@point;
+		push @values, $value;
+	}
+
+	$self->{test_points} = \@points;
+	$self->{test_values} = \@values;
+
+	return $self;
+}
+
+sub typeRef { return Value::Type('Boolean') }
+
+package context::Boolean::BOP;
+our @ISA = qw(Parser::BOP);
+
+sub _check {
+	my $self = shift;
+	# constants don't have type methods
+	# TODO: should that be "fixed"?
+	my $lType = $self->{lop}->type || $self->{lop}{type};
+	my $rType = $self->{rop}->type || $self->{lop}{type};
+
+	$self->{type} = Value::Type('Boolean');
+}
+
+sub perl {
+	my $self   = shift;
+	my $result = $self->SUPER::perl(@_);
+	my $pkg    = $self->Package('Real');
+	return "$pkg->new($result ? 1 : 0)";
+}
+
+package context::Boolean::BOP::or;
+our @ISA = qw(context::Boolean::BOP);
+
+sub _eval { return $_[1] || $_[2] ? 1 : 0 }
+
+package context::Boolean::BOP::and;
+our @ISA = qw(context::Boolean::BOP);
+
+sub _eval { return $_[1] && $_[2] ? 1 : 0 }
+
+package context::Boolean::UOP::not;
+our @ISA = qw(Parser::UOP);
+
+sub _check {
+	my $self  = shift;
+	my $lType = $self->{op}->type || $self->{op}{type};
+
+	$self->{type} = Value::Type('Boolean');
+}
+
+sub _eval { !($_[1]) ? 1 : 0 }
+
+sub perl {
+	my $self   = shift;
+	my $result = $self->SUPER::perl(@_);
+	my $pkg    = $self->Package('Real');
+	return "$pkg->new($result ? 1 : 0)";
+}
+
+package context::Boolean::BOP::xor;
+our @ISA = qw(context::Boolean::BOP);
+
+sub _eval { return !($_[1]) != !($_[2]) ? 1 : 0 }
+
+package context::Boolean::Value;
+our @ISA = qw(Value::Real);
+
+# subclass is necessary because Parser::Constant::getValueType attempts
+# to dereference the incoming value as an ARRAY -- unclear why that is
+sub getValueType {
+	# note value & equation are reversed from Value::getValueType when subclassed
+	my $value    = shift;
+	my $equation = shift;
+	return ($value,                               $value->typeRef) if ref($value) eq 'context::Boolean::Value';
+	return (context::Boolean::Value->new($value), Value::Type('Boolean'));
+}
+
+sub isBoolean {1}
+sub getType   {'Boolean'}
+sub typeRef   { Value::Type('Boolean') }
+sub class     {'Value'}
+sub type      {'Boolean'}
```

Comment:
You are working too hard, here, to coerce the types and class, and not doing the needed work to process the boolean values.  I have recommended calling this `context::Boolean::Boolean` so that the class is `Boolean` automatically, and the type remains `Number`, as it should.

Instead, you should have the constructors for this object (the `new` and `make` methods) do the checking that the number that it gets is either 1 or 0, and report an error otherwise.  And you should make its `string` and `TeX` functions print the correct things.  That's really it.  Here is my recommendation:

``` perl
package context::Boolean::Boolean;
our @ISA = ('Value::Real');

sub new {
  my $self = shift;
  my $value = $self->SUPER::new(@_);
  $value->checkBoolean unless $value->classMatch("Formula");
  return $value;
}

sub make {
  my $self = shift;
  my $result = $self->SUPER::make(@_);
  $result->checkBoolean unless $result->classMatch("Formula");
  return $result;
}

sub checkBoolean {
  my $self = shift;
  $self->Error("Numeric values can only be 1 or 0 in this context")
    unless $self->value == 1 || $self->value == 0;
}

sub string {
  my $self = shift;
  return ('F', 'T')[$self->value];
}

sub TeX {
  my $self = shift;
  return ('\bot ', '\top ')[$self->value];
}
```

Since this makes a new class `Boolean`, you may want to add

``` perl
$context->{precedence}{Boolean} = $context->{precedence}{Real};
```

to the `Init()` function so that object promotion will be done properly for `Boolean` objects.